### PR TITLE
EID-1105: Update to Java 11

### PR DESCRIPTION
**Pull request by tomrosier**
This is working in Java 11, but we cannot merge as it will break the existing Jenkins workflow. There have been some changes to the way `SunPKCS11` has been implemented between Java 8 & 11, which means that implementations are not forwards or backwards compatible. 

It is possible to use a multiple version jar, which would use 2 different classes for the same functionality but this adds extra complexity. We need to clarify if there Jenkins boxes will be upgraded from Java 8 -> 11 before continuing.

Acceptance criteria

- [x] Update JDK version to 11
- [x] Upgrade Gradle for version that supports Java 11 (at least 4.10)
- [x] Upgrade Project dependencies to support Java 11
- [x] Upgrade build scripts for Java 11 support
- [x] Update Travis config to run builds on JDK11
- [x] Check project builds locally without errors using JDK11
- [x] Update dockerfile to use jdk11 for testing and jdk11-jre for runtime
- [x] Update IntelliJ config to use JDK11
- [ ] Update the spreadsheet with progress

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -1,19 +1,19 @@
-FROM gradle:4.7.0-jdk8 as build
+FROM gradle:5.4.1-jdk11 as build
 
 WORKDIR /trust-anchor
 USER root
 ENV GRADLE_USER_HOME ~/.gradle
 
 COPY build.gradle build.gradle
-RUN gradle install
+RUN gradle installDist
```
Comment (michaelabenyohai) [inline Dockerfile]: I'm probably missing something but why is `gradle installDist` called twice?

Comment (tomrosier) [reply Dockerfile]: Good spot, no this is a mistake

**Comment by tomrosier** [comment]
`Security.addProvider(p); `... 😂

File: src/main/java/uk/gov/ida/eidas/utils/keyloader/PKCS11KeyLoader.java [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+          NoSuchAlgorithmException,
+          IOException,
+          KeyStoreException {
+
+    final KeyStore keyStore = KeyStore.getInstance("PKCS11", this.provider);
+    keyStore.load(null, keyPassword.toCharArray());
+
+    Certificate certificate = keyStore.getCertificate(alias);
+    if (certificate.equals(null)) {
+      throw new RuntimeException(("Certificate not found with alias: " + alias));
```
Comment (JHjava) [inline src/main/java/uk/gov/ida/eidas/utils/keyloader/PKCS11KeyLoader.java]: Are there any more specific exceptions we could use?

File: src/main/java/uk/gov/ida/eidas/utils/keyloader/PKCS11KeyLoader.java [hunk 3]
```diff
@@ -0,0 +1,60 @@
+package uk.gov.ida.eidas.utils.keyloader;
+
+import java.io.IOException;
+import java.security.*;
```
Comment (JHjava) [inline src/main/java/uk/gov/ida/eidas/utils/keyloader/PKCS11KeyLoader.java]: Let's avoid the wildcard imports if poss.

**Comment by JHjava** [comment]
Looks great Tom, just some slight picky points.

**Comment by JHjava** [comment]
Looks great Tom, just some slight picky points.

File: src/main/java/uk/gov/ida/eidas/utils/keyloader/PKCS11KeyLoader.java [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+          SecurityException,
+          IllegalArgumentException,
+          IOException,
+          CertificateException {
+    final KeyStore keyStore = KeyStore.getInstance("PKCS11", this.provider);
+    keyStore.load(null, keyPassword.toCharArray());
+    final PrivateKeyEntry keyEntry = (PrivateKeyEntry)keyStore.getEntry(keyAlias,
+            new KeyStore.PasswordProtection(keyPassword.toCharArray()));
+    if (keyEntry == null) {
+      throw new KeyStoreException("Key store contains wrong kind of credential, need Private key");
```
Comment (CharlesIC) [inline src/main/java/uk/gov/ida/eidas/utils/keyloader/PKCS11KeyLoader.java]: I've only notices now - this file is indented with 2 spaces but our style guide mandates 4
https://gds-way.cloudapps.digital/manuals/programming-languages/java.html#java-style-guide

Comment (tomrosier) [reply src/main/java/uk/gov/ida/eidas/utils/keyloader/PKCS11KeyLoader.java]: The whole project is only indented with 2 spaces.

Comment (CharlesIC) [reply src/main/java/uk/gov/ida/eidas/utils/keyloader/PKCS11KeyLoader.java]: Nope - other files are indented with 4. IntelliJ shows indentation in the bottom right-hand corner

Comment (CharlesIC) [reply src/main/java/uk/gov/ida/eidas/utils/keyloader/PKCS11KeyLoader.java]: `SignWithFile` and `SignWithSmartcard` are also indented with 2 spaces - they should be re-indented with 4 to match the rest of the codebase

Comment (tomrosier) [reply src/main/java/uk/gov/ida/eidas/utils/keyloader/PKCS11KeyLoader.java]: Done

**Pull request merged**